http/client.go

```diff
@@ -272,28 +276,52 @@ func (c *Client) sendRequest(
 	return rawResp, err
 }
 
+func seekable(r io.Reader, length int64) (io.ReadSeeker, error) {
+	if r == nil {
+		return nil, nil
+	}
+	if r, ok := r.(io.ReadSeeker); ok {
+		return r, nil
+	}
+	if length > maxBufSize {
+		return nil, fmt.Errorf("body of length %d is too large to hold in memory (max %d bytes)", length, maxBufSize)
+	}
+	reqData := make([]byte, int(length))
+	nrRead, err := io.ReadFull(r, reqData)
+	if err != nil {
+		return nil, errors.Newf(err, "failed reading the request data, read %v of %v bytes", nrRead, length)
+	}
+	return bytes.NewReader(reqData), nil
+}
+
 func (c *Client) sendRateLimitedRequest(
 	method, URL string,
 	headers http.Header,
-	reqData []byte,
+	reqReader io.ReadSeeker,
+	length int64,
 	logger logging.Logger,
 ) (resp *http.Response, err error) {
 	for i := 0; i < c.maxSendAttempts; i++ {
-		var reqReader io.Reader
-		if reqData != nil {
-			reqReader = bytes.NewReader(reqData)
+		var body io.ReadCloser
+		var notifier closeNotifier
+		if reqReader != nil {
+			notifier = make(closeNotifier)
+			body = struct {
+				io.Closer
+				io.Reader
+			}{notifier, reqReader}
 		}
-		req, err := http.NewRequest(method, URL, reqReader)
+		req, err := http.NewRequest(method, URL, nil)
 		if err != nil {
-			err = errors.Newf(err, "failed creating the request %s", URL)
-			return nil, err
+			return nil, errors.Newf(err, "failed creating the request %s", URL)
 		}
+		req.Body = body
```

Comment:
please add a comment stating why you're setting req.Body, rather than passing it into NewRequest

Reply:
Good catch. It turns out that it wasn't necessary in fact, as NewRequest does a dynamic type check for io.ReadCloser.

I've also added some more tests that check the new logic - there were some significant gaps in coverage previously.